File: backend/api/models/gameplay.py
# py stores algorithms pertaining to gameplay logic. there is no shared data here.

from random import shuffle

def generateRandomDeck():
    deck = [i for i in list(range(30, 154)) if i % 10 <= 3]
    shuffle(deck)
    return deck

def validPlay(play):
  cards = parseHand(play['cards'])
  table = parseHand(play['table']) if play['table'] else None
  
  if table == None and cards:
    return True
  elif cards == None:
    return False
  else:
    return cards['combo'] == table['combo'] and cards['power'] > table['power']

def parseHand(hand):
  # take hand like [140, 141, 142, 143] i.e. [A♦, A♣, A♥, A♠], and return { combo: '5x', power: 143 }, null if invalid
  
  ranks = [i // 10 for i in hand] # rank only i.e. [14, 14, 14, 14]
  suits = [i % 10 for i in hand] # suit only i.e. [0, 1, 2, 3]

  if len(hand) == 1:
    return { 'combo': '1x', 'power': hand[0] }
  elif len(hand) == 2 and allEqual(ranks):
    return { 'combo': '2x', 'power': hand[1] }
  elif len(hand) == 3 and allEqual(ranks):
    return { 'combo': '3x', 'power': hand[2] }
  elif len(hand) == 5:
    power = pokerPower(hand, ranks, suits)
    return { 'combo': '5x', 'power': power } if power else None
  else:
    return None
  
def pokerPower(hand, ranks, suits):
  # Inputs: hand [120, 121] ranks [12, 12] suits [1, 2]
  # Output: power
  # 1. straight: +1000 power (JQKA2 is highest, 23456 is lowest)
  # 2. flush: +2000 power
  # 3. full house: +3000 power
  # 4. four of a kind + 1: +4000 power
  # 5. straight flush: +5000 power

  straight = allConsecutive(ranks)
  flush = allEqual(suits)

  if straight and flush:
    return 5000 + (hand[3] if ranks[4] == 15 and ranks[3] == 14 else hand[4]) # (JQKA2 is highest, 23456 is lowest)
  elif allEqual(ranks[:4]) or allEqual(ranks[1:]):
    return 4000 + (ranks[0] if allEqual(ranks[:4]) else ranks[4])
  elif (allEqual(ranks[:3]) and allEqual(ranks[3:])) or (allEqual(ranks[:2]) and allEqual(ranks[2:])):
    return 3000 + (hand[2] if allEqual(ranks[:3]) else hand[4])
  elif flush:
    return 2000 + ranks[4]
  elif straight:
    return 1000 + (hand[3] if ranks[4] == 15 and ranks[3] == 14 else hand[4]) # (JQKA2 is highest, 23456 is lowest)
  else:
    return None

def allEqual(array):
  for i in array:
    if i != array[0]:
      return False
  return True

def allConsecutive(array):
  for i in range(len(array) - 1):
    if array[i] + 1 != array[i + 1]:
      return False
  return True

def rankCount(card, array):
  return len([i for i in array if i // 10 == card // 10])
  
def allCombinations(a, size):
  # given an (a)rray of cards, return all combinations with 'size' length (not permutations)
  # allCombinations([1, 2, 3], 2) => [[1, 2], [1, 3], [2, 3]]
  # CONSTANT TIME for 2 <= a <= 4 !!

  if len(a) <= 2:
    if size == 2:
      return [a]
  elif len(a) == 3:
    if size == 2:
      return [[a[0], a[1]], [a[0], a[2]], [a[1], a[2]]]
    elif size == 3:
      return [a]
  elif len(a) == 4:
    if size == 2:
      return [[a[0], a[1]], [a[0], a[2]], [a[0], a[3]], [a[1], a[2]], [a[1], a[3]], [a[2], a[3]]]
    elif size == 3:
      return [[a[0], a[1], a[2]], [a[0], a[1], a[3]], [a[0], a[2], a[3]], [a[1], a[2], a[3]]]
    elif size == 4:
      return [a]
  else:
    return []

def all1x2x3x4x(hand, x):
  # get all possible singles, pairs, triplets, or fours (x = 1, 2, 3, 4)

  ret = []

  if x == 1:
    ret = [[i] for i in hand]
  else:
    # 1. filter out singles, pairs, if necessary.
    filter_unplayables = [i for i in hand if rankCount(i, hand) >= x]
    if filter_unplayables == []: # end immediately if no playable hands
      return []
    # 2. partition by rank [[x, x], [x, x]]
    partition_ranks = [[filter_unplayables[0]]]
    for i in range(1, len(filter_unplayables)):
      if filter_unplayables[i] // 10 == partition_ranks[len(partition_ranks) - 1][0] // 10:
        partition_ranks[len(partition_ranks) - 1].append(filter_unplayables[i])
      else:
        partition_ranks.append([filter_unplayables[i]])
    # 3. add all perms in each rank division
    for partition in partition_ranks:
      ret += allCombinations(partition, x)
  # 4. finish

  return ret if x == 4 else sorted(ret, key = lambda i: parseHand(i)['power'])

def all5x(hand, x):
  # get straight flushes, four of a kinds, full houses, flushes, and straights

  if len(hand) < 5: # end immediately if no playable hands
    return []

  def rank(card):
    return card // 10
  def suit(card):
    return card % 10

  ret = []
  ranks = [rank(i) for i in hand]

  if x == '4x':
    # 1. get all fours
    fours = all1x2x3x4x(hand, 4) # like [[x, x, x, x]]
    # 2. for each four, add a single of every other rank
    for four in fours:
      for i in hand:
        if rank(i) != rank(four[0]):
          ret.append(four + [i])
  elif x == 'full house':
    # 1. get all pairs and triplets
    pairs = all1x2x3x4x(hand, 2)
    triplets = all1x2x3x4x(hand, 3)
    # 2. for each pair, add a triplet of every other rank
    for pair in pairs:
      for trip in triplets:
        if rank(trip[0]) != rank(pair[0]):
          ret.append(pair + trip)
  else:
    # generate flushes and straights without returning. the intersection of these will be straight flushes
    # 1. generate all flushes with unique max rank with lowest 4 as fodder
    flushes = []
    for _suit in range(4):
      filter_suit = [i for i in hand if suit(i) == _suit]
      if len(filter_suit) < 5:
        continue
      for flush_max_card in filter_suit[4:]:
        flushes.append(filter_suit[:4] + [flush_max_card])
    # 2. generate all straights
    straights = []
    for i in range(len(hand)):
      if (rank(hand[i]) + 1 in ranks) and (rank(hand[i]) + 2 in ranks) and (rank(hand[i]) + 3 in ranks) and (rank(hand[i]) + 4 in ranks):
        new_straights = [[hand[i]]]
        additions = []
        for next_rank in [1, 2, 3, 4]:
          all_next_rank = [j for j in hand if rank(j) == ranks[i] + next_rank]
          for next_rank_card in all_next_rank:
            for straight in new_straights:
              additions.append(straight + [next_rank_card])
          new_straights = additions
          additions = []
        straights += [straight for straight in new_straights if len(straight) == 5]
    # 3. generate all straight flushes
    # note: can't do [straight for straight in straights if straight in flushes] because flushes doesn't contain every single flush
    straight_flushes = [straight for straight in straights if allEqual([suit(i) for i in straight])]
    # 4. prepare to return
    if x == 'straight':
      ret = [straight for straight in straights if straight not in straight_flushes]
    elif x == 'flush':
      ret = [flush for flush in flushes if flush not in straight_flushes]
    elif x == 'straight flush':
      ret = straight_flushes
  # 4. finish
  return sorted(ret, key = lambda i: parseHand(i)['power'])

def possibilities(hand, table):
  # given a hand and a table, return all possible valid combinations to play to the table

  ret = []
  table = parseHand(table)

  # 1. add all possibilities (algorithm generates no duplicates)
  if not table:
    # what hands the AI will play first is determined in the weight() method.
    # the order is as below, although the below has no bearing on the actual order.
    ret += all5x(hand, 'straight')
    ret += all1x2x3x4x(hand, 2)
    ret += all1x2x3x4x(hand, 1)
    ret += all1x2x3x4x(hand, 3)
    ret += all5x(hand, 'full house')
    ret += all5x(hand, 'flush')
    ret += all5x(hand, '4x')
    ret += all5x(hand, 'straight flush')
    return ret
  elif table['combo'] in ['1x', '2x', '3x']:
    return [i for i in all1x2x3x4x(hand, int(table['combo'][0])) if parseHand(i)['power'] > table['power']]
  elif table['combo'] == '5x':
    if table['power'] < 2000:
      ret += all5x(hand, 'straight')
    if table['power'] < 3000:
      ret += all5x(hand, 'flush')
    if table['power'] < 4000:
      ret += all5x(hand, 'full house')
    if table['power'] < 5000:
      ret += all5x(hand, '4x')
    if table['power'] < 6000:
      ret += all5x(hand, 'straight flush') # ai will always add straight flush.
    return [i for i in ret if parseHand(i)['power'] > table['power']]

File: models/gameplay.js
const _ = require('lodash');

// gameplay.js stores algorithms pertaining to gameplay logic. there is no shared data here.

const generateRandomDeck = () => _.shuffle(_.range(30, 154).filter(card => card % 10 <= 3));

const validPlay = (play) => {
  let cards = parseHand(play.cards),
    table = play.table ? parseHand(play.table) : null;

  if (!table && cards) {
    return true;
  } else if (!cards) {
    return false;
  } else {
    return cards.combo === table.combo && cards.power > table.power;
  }
};

const parseHand = (hand) => {
  // take hand like [140, 141, 142, 143] i.e. [A♦, A♣, A♥, A♠], and return { combo: '5x', power: 143 }, null if invalid

  const ranks = hand.map(card => ~~(card / 10)), // rank only i.e. [14, 14, 14, 14]
    suits = hand.map(card => card % 10); // suit only i.e. [0, 1, 2, 3]

  if (hand.length === 1) {
    return { combo: '1x', power: hand[0] };
  } else if (hand.length === 2 && allEqual(ranks)) {
    return { combo: '2x', power: hand[1] };
  } else if (hand.length === 3 && allEqual(ranks)) {
    return { combo: '3x', power: hand[2] };
  } else if (hand.length === 5) {
    const power = pokerPower(hand, ranks, suits);
    return power ? { combo: '5x', power } : null;
  } else {
    return null;
  }
};

const pokerPower = (hand, ranks, suits) => {
  // Inputs: hand [120, 121] ranks [12, 12] suits [1, 2]
  // Output: power
  // 1. straight: +1000 power (JQKA2 is highest, 23456 is lowest)
  // 2. flush: +2000 power
  // 3. full house: +3000 power
  // 4. four of a kind + 1: +4000 power
  // 5. straight flush: +5000 power

  const straight = allConsecutive(ranks),
    flush = allEqual(suits);

  if (straight && flush) {
    return 5000 + (ranks[4] === 15 && ranks[3] === 14 ? hand[3] : hand[4]) // JQKA2 is highest, 23456 is lowest
  } else if (allEqual(ranks.slice(0, 4) || allEqual(ranks.slice(1)))) {
    return 4000 + (allEqual(ranks.slice(0, 4))) ? ranks[0] : ranks[4];
  } else if (
      allEqual(ranks.slice(0, 3)) && allEqual(ranks.slice(3))
      || allEqual(ranks.slice(0, 2)) && allEqual(ranks.slice(2))
  ) {
    return 3000 + (allEqual(ranks.slice(0, 3)) ? hand[2] : hand[4]);
  } else if (flush) {
    return 2000 + ranks[4];
  } else if (straight) {
    return 1000 + (ranks[4] === 15 && ranks[3] === 14 ? hand[3] : hand[4]) // JQKA2 is highest, 23456 is lowest
  } else {
    return null;
  }
};

const allEqual = (array) => {
  for (let i = 0; i < array.length; i++) {
    if (array[i] != array[0]) return false;
  }
  return true;
};

const allConsecutive = (array) => {
  for (let i = 0; i < array.length - 1; i++) {
    if (array[i] + 1 != array[i + 1]) return false;
  }
  return true;
};

const rankCount = (card, array) => array.filter(item => ~~(item / 10) === ~~(card / 10)).length;

const allCombinations = (a, size) => {
  // given an (a)rray of cards, return all combinations with 'size' length (not permutations)
  // allCombinations([1, 2, 3], 2) => [[1, 2], [1, 3], [2, 3]]
  // CONSTANT TIME for 2 <= a <= 4 !!
  
  if (a.length <= 2) {
    if (size === 2) return [a]
  } else if (a.length === 3) {
    if (size === 2) return [[a[0], a[1]], [a[0], a[2]], [a[1], a[2]]];
    if (size === 3) return [a];
  } else if (a.length === 4) {
    if (size === 2) return [[a[0], a[1]], [a[0], a[2]], [a[0], a[3]], [a[1], a[2]], [a[1], a[3]], [a[2], a[3]]];
    if (size === 3) return [[a[0], a[1], a[2]], [a[0], a[1], a[3]], [a[0], a[2], a[3]], [a[1], a[2], a[3]]];
    if (size === 4) return [a];
  } else {
    return [];
  }
};

const all1x2x3x4x = (hand, x) => {
  // get all possible singles, pairs, triplets, or fours (x = 1, 2, 3, 4)

  let ret = [];

  if (x === 1) {
    ret = hand.map(card => [card]);
  } else {
    // 1. filter out singles, pairs, if necessary.
    const filterUnplayables = hand.filter(card => rankCount(card, hand) >= x);
    if (_.isEmpty(filterUnplayables)) return []; // end immediately if no playable hands
    // 2. partition by rank [[x, x], [x, x]]
    const partitionRanks = [[filterUnplayables[0]]];
    for (let i = 1; i < filterUnplayables.length; i++) {
      if (~~(filterUnplayables[i] / 10) === ~~(partitionRanks[partitionRanks.length - 1][0] / 10)) {
        partitionRanks[partitionRanks.length - 1].push(filterUnplayables[i]);
      } else {
        partitionRanks.push([filterUnplayables[i]]);
      }
    }
    // 3. add all perms in each rank division
    partitionRanks.forEach(partition => {
      ret = ret.concat(allCombinations(partition, x));
    });
  }
  // 4. finish
  return x === 4 ? ret : ret.sort((a, b) => parseHand(a).power - parseHand(b).power);
};

const all5x = (hand, x) => {
  // get straight flushes, four of a kinds, full houses, flushes, and straights

  if (hand.length < 5) return []; // end immediately if no playable hands

  const rank = (card) => ~~(card / 10);
  const suit = (card) => card % 10;

  let ret = [],
    ranks = hand.map(card => rank(card));

  if (x === '4x') {
    // 1. get all fours
    const fours = all1x2x3x4x(hand, 4); // like [[x, x, x, x]]
    // 2. for each four, add a single of every other rank
    fours.forEach(four => {
      hand.forEach(card => {
        if (rank(card) !== rank(four[0])) ret.push(four.concat([card]));
      });
    });
  } else if (x === 'full house') {
    // 1. get all pairs and triplets
    const pairs = all1x2x3x4x(hand, 2),
      triplets = all1x2x3x4x(hand, 3);
    // 2. for each pair, add a triplet of every other rank
    pairs.forEach(pair => {
      triplets.forEach(trip => {
        if (rank(trip[0]) !== rank(pair[0])) ret.push(pair.concat(trip));
      });
    });
  } else {
    // generate flushes and straights without returning. the intersection of these will be straight flushes
    // 1. generate all flushes with unique max rank with lowest 4 as fodder
    const flushes = [];
    [0, 1, 2, 3].forEach(_suit => {
      const filterSuit = hand.filter(card => suit(card) === _suit);
      if (filterSuit.length < 5) return;
      filterSuit.slice(4).forEach(flushMaxCard => {
        flushes.push(filterSuit.slice(0, 4).concat([flushMaxCard]));
      });
    });
    // 2. generate all straights
    let straights = [];
    for (let i = 0; i < hand.length; i++) {
      if (
        ranks.includes(rank(hand[i]) + 1)
        && ranks.includes(rank(hand[i]) + 2)
        && ranks.includes(rank(hand[i]) + 3)
        && ranks.includes(rank(hand[i]) + 4)
      ) {
        let newStraights = [[hand[i]]],
          additions = [];
        
        [1, 2, 3, 4].forEach(nextRank => {
          const allNextRank = hand.filter(card => rank(card) === ranks[i] + nextRank);
          allNextRank.forEach(nextRankCard => {
            newStraights.forEach(straight => {
              additions.push(straight.concat([nextRankCard]));
            });
          });
          newStraights = additions;
          additions = [];
        });
        straights = straights.concat(newStraights.filter(straight => straight.length === 5));
      }
    }
    // 3. generate all straight flushes
    // note: can't do straights.map(straight => flushes.includes(straight))because flushes doesn't contain every single flush
    const straightFlushes = straights.filter(straight => allEqual(straight.map(card => suit(card))));
    // 4. prepare to return
    if (x === 'straight') {
      ret = straights.filter(straight => !JSON.stringify(straightFlushes).includes(JSON.stringify(straight)));
    } else if (x === 'flush') {
      ret = flushes.filter(flush => !JSON.stringify(straightFlushes).includes(JSON.stringify(flush)));
    } else if (x === 'straight flush') {
      ret = straightFlushes;
    }
  }
  // 5. finish
  return ret.sort((a, b) => parseHand(a).power - parseHand(b).power);
};

const possibilities = (hand, table) => {
  // given a hand and a table, return all possible valid combinations to play to the table

  let ret = [];
  table = parseHand(table);

  // 1. add all possibilities (algorithm generates no duplicates)
  if (_.isEmpty(table)) {
    // what hands the AI will play first is determined in the weight() method.
    // the order is as below, although the below has no bearing on the actual order.
    ret = ret.concat(all5x(hand, 'straight'));
    ret = ret.concat(all1x2x3x4x(hand, 2));
    ret = ret.concat(all1x2x3x4x(hand, 1));
    ret = ret.concat(all1x2x3x4x(hand, 3));
    ret = ret.concat(all5x(hand, 'full house'));
    ret = ret.concat(all5x(hand, 'flush'));
    ret = ret.concat(all5x(hand, '4x'));
    ret = ret.concat(all5x(hand, 'straight flush'));
    return ret;
  } else if (['1x', '2x', '3x'].includes(table.combo)) {
    return all1x2x3x4x(hand, parseInt(table.combo[0])).filter(hand => parseHand(hand).power > table.power);
  } else if (table.combo === '5x') {
    if (table.power < 2000) ret = ret.concat(all5x(hand, 'straight'));
    if (table.power < 3000) ret = ret.concat(all5x(hand, 'flush'));
    if (table.power < 4000) ret = ret.concat(all5x(hand, 'full house'));
    if (table.power < 5000) ret = ret.concat(all5x(hand, '4x'));
    if (table.power < 6000) ret = ret.concat(all5x(hand, 'straight flush')); // ai will always add straight flush.
    return ret.filter(hand => parseHand(hand).power > table.power);
  }
};


module.exports = {
  generateRandomDeck,
  validPlay,
  parseHand,
  pokerPower,
  allEqual,
  allConsecutive,
  rankCount,
  allCombinations,
  all1x2x3x4x,
  all5x,
  possibilities
};
